src/vm/dllimport.cpp

```diff
@@ -6018,6 +6018,54 @@ HMODULE NDirect::LoadFromNativeDllSearchDirectories(AppDomain* pDomain, LPCWSTR
     return hmod;
 }
 
+static const int MAX_LIBNAME_VARIATIONS = 4;
+
+static void AddLibNameVariation(const char** libNameVariations, int* numberOfVariations, const char* const variation)
+{
+    assert(*numberOfVariations < MAX_LIBNAME_VARIATIONS);
+    libNameVariations[*numberOfVariations] = variation;
+    *numberOfVariations = *numberOfVariations + 1;
+}
+
+static void DetermineLibNameVariations(const char** libNameVariations, int* numberOfVariations, const SString& libName, bool libNameIsRelativePath)
+{
+    *numberOfVariations = 0;
+#ifdef FEATURE_PAL
+    // P/Invokes are often declared with variations on the actual library name.
+    // For example, it's common to leave off the extension/suffix of the library
+    // even if it has one, or to leave off a prefix like "lib" even if it has one
+    // (both of these are typically done to smooth over cross-platform differences). 
+    // We try to dlopen with such variations on the original.
+    if (!libNameIsRelativePath)
+    {
+        AddLibNameVariation(libNameVariations, numberOfVariations, "%.0s%s");       // name
+    }
+    else
+    {
+        // We check if the suffix is contained in the name, because on Linux it is common to append
+        // a version number to the library name (e.g. 'libicuuc.so.57').
+        SString::CIterator begin = libName.Begin();
+        bool containsSuffix = libName.FindASCII(begin, PAL_SHLIB_SUFFIX);
```

Comment:
It would be good to make this check more thorough by adding check that there is a `.` or end of string after the PAL_SHLIB_SUFFIX to prevent false positive in name like `libtest.sound`.